### DISCUSSION TITLE
RuCore / Системы именования / ФИАС

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#242] [Минимальный набор артефактов]
//--------------------------------------
//страница: Руководство RuCore / Системы именования / ФИАС
//--------------------------------------

```
Instance:   core-ns-fias-aoguid
InstanceOf: NamingSystem
Usage: #definition

* name =   "Core_Ns_Fias_Aoguid"

* status = #active

* kind = #identifier

* date = "2023-10-23"

* uniqueId
  * type = #uri
  * value = "urn:hl7-ru:fias:aoguid"
  * preferred = true



Instance:   core-ns-fias-houseguid
InstanceOf: NamingSystem
Usage: #definition

* name =   "Core_Ns_Fias_Houseguid"

* status = #active

* kind = #identifier

* date = "2023-10-23"

* uniqueId
  * type = #uri
  * value = "urn:hl7-ru:fias:houseguid"
  * preferred = true
```

Пример использования данных систем: https://github.com/fhir-ru/core/discussions/237
Страница с методическими рекомендациями по описанию NamingSystem: https://github.com/fhir-ru/core/discussions/245

**Comment by Vrodvi** [comment]
1. Идентификатор пространства не обозначен как предпочтительный, он не обозначен в профиле расширения и он не используется в экземплярах


Один из идентификаторов пространства следует пометить как предпочтительный:
![image](https://github.com/fhir-ru/core/assets/92928137/37fa81ee-51ad-46a8-b064-5c8bb4fd62b9)
И далее в профиле расширения и в экземпляре использовать указанный предпочтительный идентификатор просмотранства.
Сейчас в профиле расширения идентификатор пространства вообще не обозначен:
![image](https://github.com/fhir-ru/core/assets/92928137/a5a5287b-49de-4e22-84e9-ad380445aa13)
В экземпляре также идентификатор пространства пропущен:
![image](https://github.com/fhir-ru/core/assets/92928137/8130c11a-30ad-4de7-a43a-bc2182a7633c)

**Reply by eugenekogan-spb** [reply]
Предпочтительный указан, идентификаторы пространств в расширениях и экземплярах включены.

**Reply by Vrodvi** [reply]
Замечание "1." закрыто.

**Comment by Vrodvi** [comment]
2. Сколько и каких идентификаторов необходимо завести для ФИАС'а?

Один из вариантов - завести идентификаторы под каждый фрагмент идентификации из ФИАС - отдельно на адресные объекты, отдельно на дома и так далее.

Альтернативным вариантом может быть один единственный идентификатор на весь ФИАС, как сейчас это указано ("https://fias.nalog.ru/"), но в этом случае возникает вопрос - как можно различить идентификаторы разных типов (идентификатор адресного объекта от идентификатора дома, к примеру), если все они в одном пространстве? Имеется ли гарантия, что идентификаторы разных типов не пересекаются между собой?
Если идентификатор на весь ФИАС один, то почему он взят не такой, какой объявлен в СЭМД'ах?
![image](https://github.com/fhir-ru/core/assets/92928137/ea32726a-e284-4e20-bbf9-c73c048ffcb1)


Моё мнение на данный вопрос - если ФИАС использует GUIDы, то это уже и есть пространство идентификаторов (оно обозначается зарезервированным значением "urn:ietf:rfc:3986") и описывать новых пространств для ФИАС'а не нужно.
![image](https://github.com/fhir-ru/core/assets/92928137/1ccf7bae-fb13-46e8-9fac-82a465ce83e2)

**Reply by eugenekogan-spb** [reply]
Исправлено в варианте двух пространств.

**Reply by Vrodvi** [reply]
Замечание "2." закрыто.

**Comment by Vrodvi** [comment]
3. В представленных фрагментах кода (FSH) сбивается форматирование 

К примеру символ звёздочки превратился в обозначение элемента списка.
Для предотвращения искажений форматирования кода просьба все кодовые фрагменты оборачивать в специальную разметку - тройной символ апострофа.
Пример:
если указать:
\```
\* фрагмент кода 
\```
, то будет корректное отображение:
```
* фрагмент кода 
```
, а не 
* фрагмент кода

**Reply by eugenekogan-spb** [reply]
3. Спасибо. Поправил.

**Reply by Vrodvi** [reply]
Замечание "3." закрыто.

**Comment by eugenekogan-spb** [comment]
@Vrodvi осталась непонятной ремарка о "urn:ietf:rfc:3986".
Предлагается заменить на "urn:hl7-ru:fias", чтобы было определено наименование как в СЭМД, и одновременно предлагается заменить на "urn:ietf:rfc:3986", потому что новое наименование вводить не требуется.

**Reply by Vrodvi** [reply]
Да, согласен, чуть не запутались.
Дело в том, что тут смешалось обсуждение идентификатора для расширения и пространства идентификаторов.

См. на примере в комментарии https://github.com/fhir-ru/core/discussions/238#discussioncomment-7403331
продублирую этот пример:
```
Extension: Fhir_ru_ext_fias 
Id: fhir-ru-ext-fias
Title: "Кодирование адреса по ФИАС"
Description: "Кодирование адреса по ФИАС."
* ^url = "urn:hl7-ru:fias"

* extension contains
    auguid 1..1 MS and
    houseguid 0..1 MS

* extension[auguid].value[x] only Identifier
* extension[auguid].valueIdentifier.system = "urn:ietf:rfc:3986"
* extension[auguid].valueIdentifier.value 1..1 MS
* extension[auguid].valueIdentifier.value ^short = "GUID"

* extension[houseguid].value[x] only Identifier
* extension[houseguid].valueIdentifier.system = "urn:ietf:rfc:3986"
* extension[houseguid].valueIdentifier.value 1..1 MS
* extension[houseguid].valueIdentifier.value ^short = "GUID"
```
Этот пример будет соответствовать структуре из СЭМДов, где обозначен идентификатор расширения (urn:hl7-ru:fias), но сами идентификаторы даются без указания пространства (т.к. там уже сказано что это GUIDы):
![image](https://github.com/fhir-ru/core/assets/92928137/2b053a46-1170-4e0f-97d4-37dc5cda9fb7)
В FHIR нужно обязательно указывать пространство идентификаторов, и для GUIDов это должен быть "urn:ietf:rfc:3986".

Это вариант, который я бы рекомендовал выбрать для лучшего соответствия СЭМДам.
Но в этом варианте мы потеряем описание артефакта "пространство идентификаторов", т.к. будет использоваться "стандартное" пространство, которое не нужно переопределять.

----


Альтернативный вариант 1:
- всё таки определяем новое пространство (но только одно), описываем его в виде NamingSystem (в качестве идентификатора можем вернуть указанный ранее "https://fias.nalog.ru/") и используем его в расширении, тогда пространство и расширение будут следующими:
```
Instance: rf-ns-fias
InstanceOf: NamingSystem
Usage: #definition
* name = "FIAS"
* status = #active
* kind = #identifier
* date = "2023-10-23"
* uniqueId.type = #uri
* uniqueId.value = "https://fias.nalog.ru/"
* uniqueId.preferred = true
```

```
Extension: Fhir_ru_ext_fias 
Id: fhir-ru-ext-fias
Title: "Кодирование адреса по ФИАС"
Description: "Кодирование адреса по ФИАС."
* ^url = "urn:hl7-ru:fias"

* extension contains
    auguid 1..1 MS and
    houseguid 0..1 MS

* extension[auguid].value[x] only Identifier
* extension[auguid].valueIdentifier.system = "https://fias.nalog.ru/"
* extension[auguid].valueIdentifier.value 1..1 MS
* extension[auguid].valueIdentifier.value ^short = "GUID адресного объекта"

* extension[houseguid].value[x] only Identifier
* extension[houseguid].valueIdentifier.system = "https://fias.nalog.ru/"
* extension[houseguid].valueIdentifier.value 1..1 MS
* extension[houseguid].valueIdentifier.value ^short = "GUID дома"
```

----

Альтернативный вариант 2:
- определяем не одно, а два новых пространства (одно для адресных объектов, другое - для домов), описываем их в виде NamingSystem (две штуки, в качестве идентификаторов можем придумать что-нибудь вроде urn:hl7-ru:fias:auguid и urn:hl7-ru:fias:houseguid) и используем их в расширении, тогда пространство и расширения будет следующими:
```
Instance: rf-ns-fias-auguid
InstanceOf: NamingSystem
Usage: #definition
* name = "FIAS"
* status = #active
* kind = #identifier
* date = "2023-10-23"
* uniqueId.type = #uri
* uniqueId.value = "urn:hl7-ru:fias:auguid"
* uniqueId.preferred = true
```

```
Instance: rf-ns-fias-houseguid
InstanceOf: NamingSystem
Usage: #definition
* name = "FIAS"
* status = #active
* kind = #identifier
* date = "2023-10-23"
* uniqueId.type = #uri
* uniqueId.value = "urn:hl7-ru:fias:houseguid"
* uniqueId.preferred = true
```

```
Extension: Fhir_ru_ext_fias 
Id: fhir-ru-ext-fias
Title: "Кодирование адреса по ФИАС"
Description: "Кодирование адреса по ФИАС."
* ^url = "urn:hl7-ru:fias"

* extension contains
    auguid 1..1 MS and
    houseguid 0..1 MS

* extension[auguid].value[x] only Identifier
* extension[auguid].valueIdentifier.system = "urn:hl7-ru:fias:auguid"
* extension[auguid].valueIdentifier.value 1..1 MS
* extension[auguid].valueIdentifier.value ^short = "GUID адресного объекта"

* extension[houseguid].value[x] only Identifier
* extension[houseguid].valueIdentifier.system = "urn:hl7-ru:fias:houseguid"
* extension[houseguid].valueIdentifier.value 1..1 MS
* extension[houseguid].valueIdentifier.value ^short = "GUID дома"
```

**Reply by eugenekogan-spb** [reply]
Определено два пространства.

**Reply by Vrodvi** [reply]
Замечание закрыто.

**Comment by Vrodvi** [comment]
# Замечание 4.

Наименование и идентификатор не соответствуют соглашениям - https://fhir.ru/support.fhir-guide.conventions#:naming-systems-naming .
По соглашениям должно быть:

```
Instance: core-ns-fias-aoguid
...
* name = "coreNsFiasAoguid"
```

```
Instance: core-ns-fias-houseguid
...
* name = "coreNsFiasHouseguid"
```

**Reply by eugenekogan-spb** [reply]
Вопрос 1: где в Соглашении написано как именовать Instance? Верно ли, что имя Instance должно соответствовать id?
<img width="741" alt="image" src="https://github.com/fhir-ru/core/assets/62525329/24354421-9437-46ea-ab35-a7262dfb0de0">

Вопрос 2. Не смог найти в соглашениях требование об указании имени lowCamelCase, нашел требование о CamelCase, по которому надо начинать с верхнего регистра:
<img width="837" alt="image" src="https://github.com/fhir-ru/core/assets/62525329/1789d882-89cd-4b35-ab3f-276ba5b9a643">
Кроме того, видимо, надо уточнить Соглашение в части именования типов артефактов, поскольку при генерации IG обнаружилось, что требуется полное наименование типа, в частности, не systems, а NamingSystem (если я правильно понял вывод из анализа ошибок публикации). В Соглашении сейчас systems: 
<img width="832" alt="image" src="https://github.com/fhir-ru/core/assets/62525329/b19fde99-6b97-426d-a130-bb4c6b8b9bc8">

**Reply by Vrodvi** [reply]
> Вопрос 1: где в Соглашении написано как именовать Instance? Верно ли, что имя Instance должно соответствовать id?

Это скрытые значения, которые ещё не описаны.

Опытным путём можно определить, что Instance - это именно id:
![image](https://github.com/fhir-ru/core/assets/92928137/0af0f2bc-1284-4d95-be14-b9abebaa5114)


В спецификации по FSH этот момент описан в - https://build.fhir.org/ig/HL7/fhir-shorthand/reference.html#item-names (сказано, что для экземпляров указывается не имя, а идентификатор, поэтому и правила формирования должны использоваться как для идентификаторов). Также есть пример с явным комментарием о том, что Instance - это id:
![image](https://github.com/fhir-ru/core/assets/92928137/e16093c4-141c-4370-802e-2d4fbf206dc9)



Желательно дополнить соглашение в разделе описания "id".




> Вопрос 2. Не смог найти в соглашениях требование об указании имени lowCamelCase, нашел требование о CamelCase, по которому надо начинать с верхнего регистра:


Требования к именам в FSH/FHIR различаются в зависимости от объекта наименования - в основном имена должны быть CamelCase, но вот для срезов сделано исключение. 
В спецификации FSH это описано тут - https://build.fhir.org/ig/HL7/fhir-shorthand/reference.html#item-names
![image](https://github.com/fhir-ru/core/assets/92928137/066d180b-f32b-4911-ac77-201c5b07908e)



> роме того, видимо, надо уточнить Соглашение в части именования типов артефактов, поскольку при генерации IG обнаружилось, что требуется полное наименование типа, в частности, не systems, а NamingSystem (если я правильно понял вывод из анализа ошибок публикации).

Да, соглашение по описанию url нужно перепроверить, дописать там что url должны быть каноническими (с пояснениями о том что это означает и какой шаблон для их формирования).

Конкретно NamingSystem - это экземпляр данных, а не профиль, поэтому url ему и не положен. 
Про ошибки публикации - не понятно, если будут пояснения, то можно обсудить отдельно.

**Reply by eugenekogan-spb** [reply]
Вопрос об ошибках публикации снят, он был связан с автоматическим формированием канонического url.

Исправлено. Внесено в рекомендации по NamingSystem.

**Reply by Vrodvi** [reply]
Замечание "4." закрыто.

**Comment by Vrodvi** [comment]
В данных системах именования нет русских описаний с пояснениями, что это вообще такое и что это имеет какое-то отношение к  ФИАС. Если не обладать скрытыми знаниями, то понять из руководства вообще не возможно - что это такое и для чего его можно/нужно использовать. Также отсутствие описания приводит к "дырками" на странице со сводкой всех артефактов руководства по реализации, см.:
![image](https://github.com/fhir-ru/core/assets/92928137/616dc7ec-2b3c-492e-bb69-a2e1eadf99f4)

**Reply by eugenekogan-spb** [reply]
Добавил
* description = "Идентификатор адресного объекта по справочнику ФИАС" 
* description = "Идентификатор дома по справочнику ФИАС"

**Reply by eugenekogan-spb** [reply]
Теперь в табличке есть описания.
<img width="715" alt="image" src="https://github.com/fhir-ru/core/assets/62525329/82a33f4c-8ed4-45d5-a4db-acc2b12b79f8">
и в расшифровке тоже
<img width="1119" alt="image" src="https://github.com/fhir-ru/core/assets/62525329/e1cc369f-592a-44ba-bdab-5401a271f6ab">

**Reply by Vrodvi** [reply]
Замечание закрыто.
